### PR TITLE
[Merged by Bors] - Reorg events

### LINKED ISSUE #2088: EventSource missing chain_reorg
## Description

EventSource has missing `chain_reorg` topics [ETH2 Events spec](https://ethereum.github.io/eth2.0-APIs/#/Events/eventstream)

## Version

Lighthouse/v1.0.4-1abc70e/x86_64-linux (stable and unstable branch)

## Present Behaviour

return `400 Bad Request` on url `/eth/v1/events?topics=chain_reorg`

## Expected Behaviour

should work as every else topis from eth2 spec (head, block, attestation, voluntary_exit, finalized_checkpoint)

## Steps to resolve

implement `chain_reorg` event [(types.rs)](https://github.com/sigp/lighthouse/blob/1abc70e815a8dd078b8958450a7516e32cc693cd/common/eth2/src/types.rs#L744)

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

Resolves #2088

## Proposed Changes

Add the `chain_reorg` SSE event topic

### CODE REVIEW

File: common/eth2/src/types.rs
Comment (paulhauner) [inline common/eth2/src/types.rs]:
```suggestion
                |e| ServerError::InvalidServerSentEvent(format!("Chain Reorg: {:?}", e)),
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs:461]:
```suggestion
    fn find_reorg_slot(
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: Thinking about the default value for `depth` when we can't/don't detect the full distance. I can think of these options:

- `0` (current)
    - Pros: 0 doesn't *really* make sense as a re-org distance, so it's kind of like a `None`.
    - Cons: It might cause naive downstream apps to think that there was actually no re-org and become faulty.
- `SLOTS_PER_HISTORICAL_ROOT`
    - Pros: it's closer to the real value than `0`.
    - Cons: Also seems like it might cause downstream apps to be faulty.
- `current_head.slot`
    - Pros: it declares that there was a re-org of the full length of the chain so a downstream app would *technically* be safe because it would check all the slots in the chain.
    - Cons: it's probably wrong and it might cause downstream apps to do a whole bunch of unnecessary work.
- `None` (change our Rust definition so this is possible)
    - Pros: is explicit that we don't know what the distance is.
    - Cons: some downstream apps might not expect this.
    
After typing this out, I'm tempted to go with the `None` approach since it will force the LH code-base to explicitly handle the case where we chose not to determine a re-org distance. I could also make an issue on the API specs repo to make this explicit in a comment regarding the `chain_reorg` event, I think we're well within our rights to decide *not* to compute this value. 
    
What are your thoughts on this?

**Review by paulhauner — CHANGES_REQUESTED** [review]
Sorry about the delay on this review. It's looking good though. I have some minor nits and also a comment about the default `depth` value.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Thinking about this more... rather than iterating backwards until a depth of `SLOTS_PER_HISTORICAL_ROOT` in the worst case, shouldn't we be iterating backwards until the finalized head?

And in the extreme scenario where the finalized head is greater than `SLOTS_PER_HISTORICAL_ROOT` slots ago, if a validator is relying entirely on server sent events, could they function properly if we do anything other than return the actual re-org depth? Or is the re-org depth only necessary to inform something like shuffling recalculation (so only matters if it's over a threshold of 2 epochs or something).

I do think if we do anything other than return the actual re-org depth, we should raise this as a API spec repo issue.

Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: > Thinking about this more... rather than iterating backwards until a depth of SLOTS_PER_HISTORICAL_ROOT in the worst case, shouldn't we be iterating backwards until the finalized head?

We could try to calculate a re-org floor via finalization, but it's a little tricky to get right (a re-org can also finalize any number of epochs) and if we're still in the `SLOTS_PER_HISTORICAL_ROOT` array then it's cheap enough to search the whole thing. In other words, I'm not sure the optimization would be worth it.

> I do think if we do anything other than return the actual re-org depth, we should raise this as a API spec repo issue.

I agree, I started the conversation here: https://github.com/ethereum/eth2.0-APIs/pull/127

**Comment by paulhauner** [comment]
Blocking this on https://github.com/ethereum/eth2.0-APIs/pull/127 for now.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: @ajsutton makes a good point here that we could use the finalized slot of the previous has as a "floor" for the re-org distance: https://github.com/ethereum/eth2.0-APIs/pull/127#issuecomment-784734706

This falls into the category of "wrong but safe".

**Comment by paulhauner** [comment]
Unblocking since we can progress. See https://github.com/sigp/lighthouse/pull/2090#discussion_r582561660

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs:461]: I realized I made this public so it was accessible from the test I wrote for it. Is that worth keeping the function public? If not I can refactor the test.

**Comment by realbigsean** [comment]
Updated according to comments, I had left one question in response to a comment. 

> @ajsutton makes a good point here that we could use the finalized slot of the previous has as a "floor" for the re-org distance

I like this!

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs:461]: Oh sorry, that's totally fine. Leave as is :)

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: This call to `rev_iter_block_roots` will clone the head state, which is a pretty expensive operation.

The alternative to this clone would be to create a custom common-descendant-search using `BeaconChain::with_head`. This has the downside of holding a read-lock on `beacon_chain.head` for the duration of the search.

However, the duration of the search will be quite short (at most 8,096 `Hash256` equality checks). Additionally, only functions that want a *write*-lock on the head will be affected by this and AFAIK the only function that wants a write-lock is `BeaconChain::find_head` (this function!).

We call `fn find_head` fairly frequently during sync (IIRC, every 128 blocks). IIRC a state clone takes ~30ms, which would add 4 mins to syncing the entire chain at this point. That's not really a big deal, but considering the impact that state clones can have on memory fragmentation, I'd be tempted to optimize out the clone.

**Review by paulhauner — COMMENTED** [review]
Looks good! Only one optimization suggestion :)

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: refactored this, let me know what you think

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I think we should use `snapshot.beacon_state.finalized_checkpoint` instead of `fork_choice`.

It's slightly more efficient, but most importantly we avoid a deadlock condition. As it stands here, we're holding the canonical head lock (via `BeaconChain::with_head`) whilst we try to obtain the fork choice lock. If we happen to do the reverse ordering somewhere else in the program (hold the fork choice lock whilst we try to obtain the head lock) then we'll end up in a lock-ordering deadlock.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
            for i in 0..state.block_roots.len() {
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: As per previous comment about implicit saturation.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: As per previous comment about implicit saturation.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: It might be nice to explicitly use `saturating_sub` or add a comment here to show that we've considered underflow.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I spent quite a while trying to find edge-cases with the genesis state and forwards/backwards reorgs but I couldn't make anything stick.

One thing I noticed is that this line contains an implicit assumption that the two blocks share the same genesis block (or else an error is returned). This is a totally reasonable assumption, but the implicit assumption without a comment looked like an attack vector. 

I even wrote another implementation during this journey, to help me understand the problem better: https://github.com/paulhauner/lighthouse/commit/f82d2b139d1daa409830661f6cb361b6ff7516b2

I'm really not convinced that my implementation is better than yours though. It's potentially more explicit, but it's more LoC.

I think I might sit on this over the weekend and come back to it with a fresh mind. I'd also be interested in the fresh eyes of @michaelsproul, if he has the time/desire.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Looking good! I left a few notes, the most notable change regarding avoiding future deadlock conditions (it's a simple change though).

I also left a rambling note about `fn find_reorg_slot`... Perhaps after I sit on it for a bit I'll have a clearer view :sweat_smile:

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: I'm a big fan of the pair of iterators in your implementation. I think it makes it way simpler conceptually. I cherry picked that commit and pushed it, let me know if you'd like me to revert it :slightly_smiling_face:

**Comment by paulhauner** [comment]
I'm down to merge this after the following are addressed:

- https://github.com/sigp/lighthouse/pull/2090/files#r619750427
- Conflicts

Will be nice to finally get this in! :sweat_smile:

**Comment by realbigsean** [comment]
> I'm down to merge this after the following are addressed:
> 
>     * https://github.com/sigp/lighthouse/pull/2090/files#r619750427
> 
>     * Conflicts
> 
> 
> Will be nice to finally get this in! sweat_smile

Addressed!

**Comment by paulhauner** [comment]
FYI there's a few tests failing here. No rush, just letting you know.

**Comment by realbigsean** [comment]
Sorry this took me a while, but fixed the test

**Review by paulhauner — APPROVED** [review]
Woo! Nothing like approving a PR from last year :sweat_smile: 

I'll hold of on bors-ing this to see if I can get a batch going.

**Pull request closed without merging**
